File: scripts/ingest-artifact-pages.py
#!/usr/bin/env python3
"""Ingest paste-form artifacts DIRECTLY from the published tool pages.

Source:  ../nikbearbrown_com/public/artifacts/*.html  (the originals)
Output:  data/catalog/artifact-pages.json

The contract: KEEP THE STRUCTURE, DROP THE PRESENTATION. The full tag
skeleton survives — h1-h6, sections' heading hierarchy, lists, tables,
code blocks — with every class, style, and attribute stripped, so the
site's stylesheet (artifact-doc.css, built on the site's CSS variables)
owns typography and color. Source-page chrome (breadcrumbs, in-page nav,
anything before the <h1>) is dropped.

Why Python: the site ships no HTML parser and the sandbox has no network
for installs; the stdlib parser is battle-tested. Run locally, commit the
JSON — same ingest-to-file pattern as the ledger.

Usage: python3 scripts/ingest-artifact-pages.py [artifacts-dir]
"""
import json
import re
import sys
from datetime import date
from html import escape
from html.parser import HTMLParser
from pathlib import Path

SRC = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("../nikbearbrown_com/public/artifacts")
OUT = Path("data/catalog/artifact-pages.json")

KEEP = {"h1", "h2", "h3", "h4", "h5", "h6", "p", "ul", "ol", "li",
        "table", "thead", "tbody", "tr", "th", "td",
        "pre", "code", "strong", "em", "b", "i", "hr", "blockquote", "br"}
SKIP = {"script", "style", "head", "nav", "footer", "svg", "iframe",
        "noscript", "form", "button", "select", "input"}
VOID = {"hr", "br"}


class Clean(HTMLParser):
    """Emit cleaned HTML: KEEP tags attribute-free, unwrap the rest."""

    def __init__(self):
        super().__init__(convert_charrefs=True)
        self.out = []
        self.skip = 0
        self.pre = 0
        self.title = ""
        self.in_title = False
        self.meta_desc = ""
        self.span_gap = False  # chip heuristic: adjacent spans get a separator

    def handle_starttag(self, tag, attrs):
        if tag == "title":
            self.in_title = True
        if tag == "meta":
            a = dict(attrs)
            if a.get("name") == "description":
                self.meta_desc = a.get("content", "")
        if tag in SKIP:
            self.skip += 1
            return
        if self.skip:
            return
        if tag == "span" and self.span_gap:
            self.out.append(" · ")
        if tag != "span":
            self.span_gap = False
        if tag in KEEP:
            self.out.append(f"<{tag}>")
            if tag == "pre":
                self.pre += 1

    def handle_endtag(self, tag):
        if tag == "title":
            self.in_title = False
        if tag in SKIP:
            self.skip = max(0, self.skip - 1)
            return
        if self.skip:
            return
        if tag == "span":
            self.span_gap = True
            return
        self.span_gap = False
        if tag in KEEP and tag not in VOID:
            if tag == "pre":
                self.pre = max(0, self.pre - 1)
            self.out.append(f"</{tag}>")

    def handle_data(self, data):
        if self.in_title:
            self.title += data
            return
        if self.skip:
            return
        if data.strip():
            self.span_gap = False
        text = data if self.pre else re.sub(r"\s+", " ", data)
        if text:
            self.out.append(escape(text, quote=False))

    def html(self):
        h = "".join(self.out)
        # drop source-page chrome: everything before the first <h1>
        i = h.find("<h1>")
        if i > 0:
            h = h[i:]
        # prune empty elements produced by unwrapped decoration
        for _ in range(3):
            h = re.sub(r"<(p|li|ul|ol|h[1-6]|strong|em|code)>\s*</\1>", "", h)
        return h.strip()


def to_text(html):
    """Cleaned HTML -> markdown-ish clipboard text (the actual paste)."""
    s = html
    for n in range(1, 7):
        s = s.replace(f"<h{n}>", "\n\n" + "#" * n + " ").replace(f"</h{n}>", "\n")
    s = re.sub(r"<pre><code>|<pre>", "\n```\n", s)
    s = re.sub(r"</code></pre>|</pre>", "\n```\n", s)
    s = s.replace("<li>", "\n- ").replace("</li>", "")
    s = s.replace("<tr>", "\n").replace("</tr>", "")
    s = re.sub(r"</t[dh]><t[dh]>", " | ", s)
    s = s.replace("<hr>", "\n\n---\n\n").replace("<br>", "\n")
    s = s.replace("<p>", "\n\n").replace("</p>", "")
    s = re.sub(r"</?[a-z0-6]+>", "", s)
    s = s.replace("&amp;", "&").replace("&lt;", "<").replace("&gt;", ">")
    return re.sub(r"\n{3,}", "\n\n", s).strip()


pages = []
for f in sorted(SRC.glob("*.html")):
    c = Clean()
    c.feed(f.read_text(errors="replace"))
    html = c.html()
    if len(html) < 500:
        print(f"  skip {f.name}: {len(html)} chars after cleaning")
        continue
    slug = re.sub(r"-(tool|reference(-v\d+)?)$", "", f.stem)
    if any(p["slug"] == slug for p in pages):
        prev = next(p for p in pages if p["slug"] == slug)
        if len(html) <= len(prev["html"]):
            continue
        pages.remove(prev)
    title = re.split(r"\s+[—–]\s+", c.title.strip())[0].strip() or f.stem
    raw = to_text(html)
    pages.append({
        "slug": slug, "title": title,
        "description": c.meta_desc.strip(),
        "chars": len(raw), "tokensApprox": round(len(raw) / 4),
        "source": f"public/artifacts/{f.name}",
        "html": html, "raw": raw,
    })
    print(f"  {slug:26s} {title:14s} h2×{html.count('<h2>')} h3×{html.count('<h3>')} "
          f"table×{html.count('<table>')} ~{round(len(raw)/4/1000,1)}k tok")

pages.sort(key=lambda p: p["slug"])
OUT.write_text(json.dumps({"generated": str(date.today()), "pages": pages}, indent=1))
print(f"\n{len(pages)} artifact pages -> {OUT}")

File: scripts/ingest-artifact-pages.mjs
/**
 * Ingest paste-form artifacts from the bear-brown-tools repo into a committed
 * catalog file, following the ingest-to-file pattern: run locally, commit the
 * JSON, the site builds from it. No runtime dependency on the sibling repo.
 *
 * Source:  ../bear-brown-tools/artifacts/*.md   (plain markdown + frontmatter)
 * Output:  data/catalog/artifact-pages.json
 *
 * Markdown -> semantic HTML happens HERE, at ingest time, because the site
 * deliberately carries no markdown dependency. Tags only — no classes, no
 * inline styles; the site's stylesheet owns presentation.
 *
 * Usage: node scripts/ingest-artifact-pages.mjs [path-to-artifacts-dir]
 */
import { readFileSync, readdirSync, writeFileSync } from 'fs'
import { join } from 'path'

const SRC = process.argv[2] ?? join(process.cwd(), '..', 'bear-brown-tools', 'artifacts')
const OUT = join(process.cwd(), 'data', 'catalog', 'artifact-pages.json')

const esc = (s) => s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
const inline = (s) => esc(s).replace(/`([^`]+)`/g, '<code>$1</code>')

function mdToHtml(md) {
  const out = []
  const lines = md.split('\n')
  let i = 0, para = [], list = null, table = null

  const flushPara = () => { if (para.length) { out.push(`<p>${inline(para.join(' '))}</p>`); para = [] } }
  const flushList = () => { if (list) { out.push(`<ul>${list.map((x) => `<li>${inline(x)}</li>`).join('')}</ul>`); list = null } }
  const flushTable = () => {
    if (table) {
      const rows = table.map((cells, r) => {
        const tag = r === 0 ? 'th' : 'td'
        return `<tr>${cells.map((c) => `<${tag}>${inline(c)}</${tag}>`).join('')}</tr>`
      })
      out.push(`<table>${rows.join('')}</table>`)
      table = null
    }
  }
  const flushAll = () => { flushPara(); flushList(); flushTable() }

  while (i < lines.length) {
    const line = lines[i]
    if (line.trim().startsWith('```')) {
      flushAll()
      const code = []
      i++
      while (i < lines.length && !lines[i].trim().startsWith('```')) { code.push(lines[i]); i++ }
      out.push(`<pre><code>${esc(code.join('\n'))}</code></pre>`)
      i++
      continue
    }
    const h = line.match(/^(#{1,6})\s+(.*)$/)
    if (h) { flushAll(); const n = h[1].length; out.push(`<h${n}>${inline(h[2])}</h${n}>`); i++; continue }
    if (/^\s*-\s+/.test(line)) { flushPara(); flushTable(); (list ??= []).push(line.replace(/^\s*-\s+/, '')); i++; continue }
    if (line.includes(' | ') && line.trim() !== '') {
      flushPara(); flushList()
      const cells = line.split(' | ').map((c) => c.trim()).filter((c, idx, a) => !(c === '' && (idx === 0 || idx === a.length - 1)))
      if (cells.length >= 2) { (table ??= []).push(cells); i++; continue }
    }
    if (line.trim() === '') { flushAll(); i++; continue }
    flushList(); flushTable()
    para.push(line.trim())
    i++
  }
  flushAll()
  return out.join('\n')
}

function frontmatter(text) {
  const m = text.match(/^---\n([\s\S]*?)\n---\n/)
  if (!m) return [{}, text]
  const fm = {}
  for (const line of m[1].split('\n')) {
    const kv = line.match(/^(\w+):\s*(.*)$/)
    if (kv) fm[kv[1]] = kv[2].replace(/^"|"$/g, '')
  }
  return [fm, text.slice(m[0].length)]
}

const pages = []
for (const f of readdirSync(SRC).filter((f) => f.endsWith('.md')).sort()) {
  const [fm, body] = frontmatter(readFileSync(join(SRC, f), 'utf-8'))
  const raw = body.trim()
  pages.push({
    slug: f.replace(/\.md$/, ''),
    title: fm.title ?? f.replace(/\.md$/, ''),
    description: fm.description ?? '',
    chars: raw.length,
    tokensApprox: Math.round(raw.length / 4),
    source: fm.source ?? '',
    html: mdToHtml(raw),
    raw,
  })
}

writeFileSync(OUT, JSON.stringify({ generated: new Date().toISOString().slice(0, 10), pages }, null, 1))
console.log(`${pages.length} artifact pages -> ${OUT}`)
